internal/controllers/core/portforward/reconciler.go

```diff
@@ -16,88 +21,75 @@ import (
+	store      store.RStore
+	kClient    k8s.Client
+	ctrlClient ctrlclient.Client
 
 	// map of PortForward object name --> running forward(s)
-	activeForwards map[string]portForwardEntry
+	activeForwards map[types.NamespacedName]portForwardEntry
 }
 
-func NewReconciler(kClient k8s.Client) *Reconciler {
+var _ store.TearDowner = &Reconciler{}
+var _ reconcile.Reconciler = &Reconciler{}
+
+func NewReconciler(store store.RStore, kClient k8s.Client, ctrlClient ctrlclient.Client) *Reconciler {
 	return &Reconciler{
+		store:          store,
 		kClient:        kClient,
-		activeForwards: make(map[string]portForwardEntry),
+		ctrlClient:     ctrlClient,
+		activeForwards: make(map[types.NamespacedName]portForwardEntry),
 	}
 }
 
-// Figure out the diff between what's in the data store and
-// what port-forwarding is currently active.
-func (r *Reconciler) diff(ctx context.Context, st store.RStore) (toStart []portForwardEntry, toShutdown []portForwardEntry) {
-	state := st.RLockState()
-	defer st.RUnlockState()
+func (r *Reconciler) SetClient(client ctrlclient.Client) {
+	r.ctrlClient = client
+}
+
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&PortForward{}).Complete(r)
+}
 
-	statePFs := state.PortForwards
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	err := r.reconcile(ctx, req.NamespacedName)
+	return ctrl.Result{}, err
+}
 
-	for name, existing := range r.activeForwards {
-		if _, onState := statePFs[name]; !onState {
-			// This port forward is no longer on the state, shut it down.
-			toShutdown = r.addToShutdown(toShutdown, existing)
-			continue
-		}
+func (r *Reconciler) reconcile(ctx context.Context, name types.NamespacedName) error {
+	pf := &PortForward{}
+	err := r.ctrlClient.Get(ctx, name, pf)
+
+	if apierrors.IsNotFound(err) || pf.ObjectMeta.DeletionTimestamp != nil {
+		// PortForward deleted in API server -- stop and remove it
+		r.stop(name)
+		return nil
 	}
 
-	for name, desired := range statePFs {
-		existing, isActive := r.activeForwards[name]
-		if isActive {
-			// We're already running this PortForward -- do we need to do anything further?
-			// NOTE(maia): we compare the ManifestName annotation so that if a user changes
-			//   just the manifest name, the PF logs will go to the correct place.
-			if equality.Semantic.DeepEqual(existing.Spec, desired.Spec) &&
-				existing.ObjectMeta.Annotations[v1alpha1.AnnotationManifest] ==
-					desired.ObjectMeta.Annotations[v1alpha1.AnnotationManifest] {
-				// Nothing has changed, nothing to do
-				continue
-			}
-
-			// There's been a change to the spec for this PortForward, so tear down the old version
-			toShutdown = r.addToShutdown(toShutdown, existing)
+	if active, ok := r.activeForwards[name]; ok {
+		if equality.Semantic.DeepEqual(active.Spec, pf.Spec) &&
+			equality.Semantic.DeepEqual(active.ObjectMeta.Annotations, pf.ObjectMeta.Annotations) {
```

Comment:
FWIW I think comparing spec + _just_ the `v1alpha1.AnnotationManifest` annotation (rather than all) from the old code is a bit less error-prone in the event that some{one,thing} starts manipulating annotations across resources.

Reply:
good call, will do